Add Swift bridging types that correspond to TextExtraction::Item and TextExtraction::ItemData

whsieh opened the pull request:
#### ebea90b9d06032b0e343f446ce705b1d6ca65094
<pre>
Add Swift bridging types that correspond to TextExtraction::Item and TextExtraction::ItemData
<a href="https://bugs.webkit.org/show_bug.cgi?id=269035">https://bugs.webkit.org/show_bug.cgi?id=269035</a>

Reviewed by Megan Gardner.

Add simple Swift wrappers around the `TextExtraction` C++ structs along with helper functions to
recursively convert a `WebCore::TextExtraction::Item` into a `WKTextExtractionItem` in Objective-C
code (which can eventually be used in Swift code as well).

See below for more details.

* Source/WebKit/SourcesCocoa.txt:

* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.h: Copied from Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h.
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.mm: Added.

Add a helper file that contains a new utility function to convert a `WebCore::TextExtraction::Item`
struct into a nested `WKTextExtractionItem`, which will eventually be used in internal Swift code.

(WebKit::containerType):
(WebKit::createItemIgnoringChildren):
(WebKit::createItemRecursive):
(WebKit::createItem):

Add logic to do the conversion above.

* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:

Add the ability to soft-link all of the new Swift/ObjC types in libWebKitSwift.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionItem.h: Added.

Create a header file that contains only interface, property and method declarations used in WebKit
to interact with the new Swift types in libWebKitSwift.

* Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift: Added.

Add the new (ObjC-capable) Swift classes to libWebKitSwift.

(WKTextExtractionItem.rectInRootView):
(WKTextExtractionItem.children):
(WKTextExtractionContainerItem.container):
(WKTextExtractionTextItem.content):
(WKTextExtractionScrollableItem.contentSize):
(WKTextExtractionEditableItem.isFocused):
(WKTextExtractionInteractiveItem.isEnabled):
(WKTextExtractionImageItem.name):
(WKTextExtractionImageItem.altText):
* Source/WebKit/WebKitSwift/WebKitSwift.h

Add `WKTextExtractionItem.h` to the umbrella header.

Canonical link: <a href="https://commits.webkit.org/274363@main">https://commits.webkit.org/274363@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a1aa188191f6bc6b1580f70a6a52e338518c75c1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38846 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17777 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41191 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41377 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34499 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20611 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15126 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32547 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39419 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14950 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33682 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12989 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12958 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42653 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35247 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34945 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38777 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13656 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11253 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36997 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15268 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8702 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/14926 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14749 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

megangardner reviewed (approved):
Except the swift part, that needs someone else's eyes...

rr-codes commented on Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift:
```diff
@@ -0,0 +1,110 @@
+
+@available(iOS 17.0, macOS 12.0, *)
+@objc(WKTextExtractionItem) public class WKTextExtractionItem: NSObject {
+    public var rectInRootView: CGRect
```
should this really be mutable?

whsieh replied:
(Ditto, w.r.t. the other comment)

rr-codes commented on Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift:
```diff
@@ -0,0 +1,110 @@
+@objc(WKTextExtractionContainerItem) public class WKTextExtractionContainerItem: WKTextExtractionItem {
+    public var container: WKTextExtractionContainer
+
+    @objc public init(container: WKTextExtractionContainer, rectInRootView: CGRect) {
```
```suggestion
    @objc public init(with container: WKTextExtractionContainer, rectInRootView: CGRect) {
```

tiny nit to be more conventionally  Swift

whsieh replied:
Adding `with` here results in a crash when attempting to initialize `WKTextExtractionContainerItem`.

However, it seems to work fine for the `WKTextExtractionItem` above so I'll do that.

rr-codes commented on Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift:
```diff
@@ -0,0 +1,110 @@
+
+@available(iOS 17.0, macOS 12.0, *)
+@objc(WKTextExtractionTextItem) public class WKTextExtractionTextItem: WKTextExtractionItem {
+    public var content: String
```
ditto about mutability

whsieh replied:
This was just mutable to match the C++ struct, but I suppose I don't really expect anything to actually `-setContent:`. I'll see if I can make it `readonly` (or, the Swift equivalent of `readonly`)

(Edit: apparently `readonly` is `let` in Swift? It seems to work if I use `let`)

rr-codes commented on Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift:
```diff
@@ -0,0 +1,110 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+
+import CoreFoundation
```
is CoreFoundation needed?

whsieh replied:
`CGRect` seems to come from CoreFoundation instead of CoreGraphics, for some reason 🤷

rr-codes replied:
does Foundation not itself import CoreFoundation though? I swear I’ve used CGRect and have never imported CoreFoundation lol

whsieh replied:
Yeah, I can try omitting it

rr-codes commented on Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift:
```diff
@@ -0,0 +1,110 @@
+import Foundation
+
+@available(iOS 17.0, macOS 12.0, *)
+@objc(WKTextExtractionItem) public class WKTextExtractionItem: NSObject {
```
I think this can just be `@objc` (you only need the label if you want the objc class to be named different than the swift class)

whsieh replied:
I think this caused soft-linking the ObjC class to fail when I omitted it, for some reason. I can try again and confirm..

whsieh replied:
Yeah, `@objc` doesn't work:

```
/Users/whsieh/Build/Release/usr/local/include/WebKitSwift/WKTextExtractionItem.h:43:12: error: declaration has external linkage, but dynamic library doesn't have symbol 'WKTextExtractionItem'
   43 | @interface WKTextExtractionItem : NSObject
…
```

Additionally, removing the declarations results in `getWKTextExtractionItemClass()` returning `nil`.

whsieh committed:
Add Swift bridging types that correspond to TextExtraction::Item and TextExtraction::ItemData
https://bugs.webkit.org/show_bug.cgi?id=269035

Reviewed by Megan Gardner.

Add simple Swift wrappers around the `TextExtraction` C++ structs along with helper functions to
recursively convert a `WebCore::TextExtraction::Item` into a `WKTextExtractionItem` in Objective-C
code (which can eventually be used in Swift code as well).

See below for more details.

* Source/WebKit/SourcesCocoa.txt:

* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.h: Copied from Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h.
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.mm: Added.

Add a helper file that contains a new utility function to convert a `WebCore::TextExtraction::Item`
struct into a nested `WKTextExtractionItem`, which will eventually be used in internal Swift code.

(WebKit::containerType):
(WebKit::createItemIgnoringChildren):
(WebKit::createItemRecursive):
(WebKit::createItem):

Add logic to do the conversion above.

* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:

Add the ability to soft-link all of the new Swift/ObjC types in libWebKitSwift.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionItem.h: Added.

Create a header file that contains only interface, property and method declarations used in WebKit
to interact with the new Swift types in libWebKitSwift.

* Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift: Added.

Add the new (ObjC-capable) Swift classes to libWebKitSwift.

(WKTextExtractionItem.rectInRootView):
(WKTextExtractionItem.children):
(WKTextExtractionContainerItem.container):
(WKTextExtractionTextItem.content):
(WKTextExtractionScrollableItem.contentSize):
(WKTextExtractionEditableItem.isFocused):
(WKTextExtractionInteractiveItem.isEnabled):
(WKTextExtractionImageItem.name):
(WKTextExtractionImageItem.altText):
* Source/WebKit/WebKitSwift/WebKitSwift.h

Add `WKTextExtractionItem.h` to the umbrella header.

Canonical link: https://commits.webkit.org/274363@main
